### PR TITLE
Redirect to an error page on 404!

### LINKED ISSUE #7658: Redirect to an error page when a local file or URL cannot be loaded.

### PR DESCRIPTION
**Pull request by ghost**
fixes #7658 

[Review on Reviewable](https://reviewable.io/reviews/servo/servo/8650)

### CODE REVIEW

**Comment by Wafflespeanut** [comment]
Since we're doing everything in the `net`, this is independent of the sub-issues mentioned there. And, it works! (Running `./mach run blaaaaaaah.html`)

not-found

**Comment by Wafflespeanut** [comment]
@jdm Should we go for returning a page for HTTP 404? If so, I'm thinking of checking the headers in http_loader and adding a field to LoadError, which we can direct to about:not-found in `load_consumer`. Thoughts?

**Comment by jdm** [comment]
My fear is that we may not be able to determine whether to show a custom error page or not at the point when we're returning a LoadError value, since there are lots of pages that (presumably) have 404 status codes but also include a body that they want us to show.

**Comment by Wafflespeanut** [comment]
Okay, I'll fix it as a part of this PR itself (we don't want this for an E-easy issue) :)

**Comment by eefriedman** [comment]
If I'm reading the HTML/fetch standards correctly, a 404 status and a 200 status are basically equivalent for the purpose of the navigation algorithm.  (IE has some special handling, https://support.microsoft.com/en-us/kb/294807 etc., but as far as I can tell other browsers don't.)

**Comment by jdm** [comment]
@bors-servo: delegate+
One small change and then we're good to go.
-S-awaiting-review +S-needs-code-changes 

---

Reviewed 4 of 4 files at r1.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[components/net/file_loader.rs, line 106 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8650#-K4-3kJw223cRnNPs3n1:-K4-3kJw223cRnNPs3n2:1041379088) ([raw file](https://github.com/servo/servo/blob/1dafbe5ecb2844526f2b267596e5f1b98585788c/components/net/file_loader.rs#L106)):*
Let's make a proper new url here; I find the resulting `file://not-found` spooky.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8650#-:-K4-3wA82szFv4EPCP73:-1402163712)*

**Comment by jdm** [comment]
@Wafflespeanut: It doesn't look my final comment about the URL was addressed?

File: components/net/file_loader.rs [hunk 1]
```diff
@@ -96,8 +98,13 @@ pub fn factory(load_data: LoadData,
...
                     }
-                    Err(e) => {
-                        send_error(url, e.description().to_owned(), senders);
+                    Err(_) => {
+                        // this should be one of the three errors listed in
+                        // http://doc.rust-lang.org/std/fs/struct.OpenOptions.html#method.open
+                        // but, we'll go for a "404 Not Found!"
+                        let url = Url::parse("about:not-found").unwrap();
+                        let load_data_404 = LoadData::new(url, None);
+                        about_loader::factory(load_data_404, senders, classifier, cancel_listener)
```
Comment (Wafflespeanut) [inline components/net/file_loader.rs]: @jdm I'm confused what went wrong. I can see my changes, but reviewable can't see it.

**Comment by Wafflespeanut** [comment]
@bors-servo: retry

- #8676

**Comment by Wafflespeanut** [comment]
@bors-servo: retry

- #8676

**Pull request merged**